lib/floe/workflow/payload_template.rb

```diff
@@ -4,31 +4,56 @@ module Floe
   class Workflow
     class PayloadTemplate
       def initialize(payload)
-        @payload = payload
+        @payload = parse_payload(payload)
       end
 
       def value(context, inputs = {})
-        interpolate_value_nested(payload, context, inputs)
+        interpolate_value(payload, context, inputs)
       end
 
       private
 
       attr_reader :payload
 
-      def interpolate_value_nested(value, context, inputs)
+      def parse_payload(value)
         case value
         when Array
-          value.map { |val| interpolate_value_nested(val, context, inputs) }
+          value.map { |val| parse_payload(val) }
         when Hash
           value.to_h do |key, val|
             if key.end_with?(".$")
-              [key.chomp(".$"), interpolate_value_nested(val, context, inputs)]
+              raise Floe::InvalidWorkflowError, "both #{key} and #{key.chomp(".$")} present" if value.keys.include?(key.chomp(".$"))
+
+              [key, parse_payload(val)]
             else
               [key, val]
             end
           end
         when String
-          value.start_with?("$") ? Path.value(value, context, inputs) : value
+          if value.start_with?("$")
+            Path.new(value)
+          else
+            value
+          end
+        else
+          value
+        end
+      end
+
+      def interpolate_value(value, context, inputs)
```

Comment:
this works as you described it. And maybe we want to merge this as is and plan a possible continuation?

Does it make sense to take it a little further and have `PayloadTemplateArray`, `PayloadTemplateHash`, `PayloadTemplateStatic`, and `Path`?

Then all methods are like a command pattern with:

```ruby
def value(context, inputs)
  @template.value(context, inputs)
end
```

Maybe the recursive nature breaks this down too quickly.
Think I do like keeping `Path`, `.$` and static values hidden from the array/hash/constant parser.

Not sure if this is overcomplicating things though.